plugins/targetlocker/dblocker/dblocker.go

```diff
@@ -0,0 +1,310 @@
+	// phantom reads might mess with the expired lock cleaning assumptions below,
+	// so request serializable isolation
+	tx, err := d.db.BeginTx(context.Background(), &sql.TxOptions{Isolation: sql.LevelSerializable})
+	if err != nil {
+		return fmt.Errorf("unable to start database transaction: %w", err)
+	}
+	defer func() {
+		// this always fails if tx.Commit() was called before, ignore error
+		_ = tx.Rollback()
+	}()
+
+	// clean expired locks first, this simplifies the logic later
+	if err := d.cleanExpired(tx, targets, now); err != nil {
+		return err
+	}
+
+	locks, err := d.queryLocks(tx, targets)
+	if err != nil {
+		return err
+	}
+	// go through existing locks, they are either held by something else (abort)
+	// help by us (update time), or not held (insert)
+	inserts := make([]string, 0)
+	updates := make([]string, 0)
+	conflicts := make([]dblock, 0)
+	for _, t := range targets {
+		lock, ok := locks[t]
+		switch {
+		case !ok:
+			inserts = append(inserts, t)
+		case lock.jobID == jobID:
+			updates = append(updates, t)
+		default:
+			conflicts = append(conflicts, lock)
+		}
+	}
+	if len(conflicts) > 0 {
+		return fmt.Errorf("unable to lock, targets have conflicting locks: %v", conflicts)
+	}
+
+	ins := "INSERT INTO locks (target_id, job_id, created_at, expires_at) VALUES (?, ?, ?, ?);"
+	for _, id := range inserts {
+		if _, err := tx.Exec(ins, id, jobID, now, expires); err != nil {
+			return fmt.Errorf("unable to lock target %s: %w", id, err)
+		}
+	}
+
+	upd := "UPDATE locks SET expires_at = ? WHERE target_id = ? AND job_id = ?;"
+	for _, id := range updates {
+		if _, err := tx.Exec(upd, expires, id, jobID); err != nil {
+			return fmt.Errorf("unable to refresh lock on target %s: %w", id, err)
+		}
+	}
+
+	return tx.Commit()
+}
+
+// handleUnlock does the real unlocking, it assumes the jobID is valid
+func (d *DBLocker) handleUnlock(jobID int64, targets []string) error {
+	// unlocking is all or nothing in one transaction
+	// phantom reads might mess with the expired lock cleaning assumptions below,
+	// so request serializable isolation
+	tx, err := d.db.BeginTx(context.Background(), &sql.TxOptions{Isolation: sql.LevelSerializable})
+	if err != nil {
+		return fmt.Errorf("unable to start database transaction: %w", err)
+	}
+	defer func() {
+		// this always fails if tx.Commit() was called before, ignore error
+		_ = tx.Rollback()
+	}()
+
+	// clean expired locks first, this simplifies the logic later
+	if err := d.cleanExpired(tx, targets, time.Now()); err != nil {
+		return err
+	}
+
+	locks, err := d.queryLocks(tx, targets)
+	if err != nil {
+		return err
+	}
+
+	// detect conflicts (unlock on foreign locks) and warn about them,
+	// but don't abort.
+	conflicts := make([]dblock, 0)
+	for _, lock := range locks {
+		if lock.jobID != jobID {
+			conflicts = append(conflicts, lock)
+		}
+	}
+	if len(conflicts) > 0 {
+		log.Warningf("unable to unlock some targets due to different lock owners: %+v", conflicts)
```

Comment:
let's add the job ID that is failing the request here too

Reply:
same as above

Reply:
same as above :P

Reply:
done, same as above :P